miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/RealMiniApp.kt

```diff
@@ -113,11 +153,56 @@ internal class RealMiniApp(
         }
     }
 
+    private fun isRequiredPermissionDenied(appId: String): Boolean {
+        getCachedRequiredPermissions(appId).find {
+            it.second != MiniAppCustomPermissionResult.ALLOWED
+        }?.let { return true }
+        return false
+    }
+
+    private fun getCachedRequiredPermissions(
+        appId: String
+    ): List<Pair<MiniAppCustomPermissionType, MiniAppCustomPermissionResult>> {
+        val requiredPermissions =
+            arrayListOf<Pair<MiniAppCustomPermissionType, MiniAppCustomPermissionResult>>()
+
+        val manifest = miniAppManifestCache.readMiniAppManifest(appId)
+        // TODO: check if it throws error for empty version id
+        manifest.requiredPermissions.forEach { (first) ->
+            miniAppCustomPermissionCache.readPermissions(appId).pairValues.find {
+                it.first == first
+            }?.let { requiredPermissions.add(it) }
+        }
+
+        return requiredPermissions
```

Comment:
Couldn't you use `mapNotNull` here and for `getCachedOptionalPermissions`?
```suggestion
        val manifest = miniAppManifestCache.readMiniAppManifest(appId)
        val cachedPermissions =  miniAppCustomPermissionCache.readPermissions(appId).pairValues
        return manifest.requiredPermissions.mapNotNull {(first) ->
            cachedPermissions.find { it.first == first }
        }
```

Reply:
I have updated this part with moving it to `MiniAppManifestCache`.